File: backends/Contracts/Implementations/Caching/Distributed/DistributedCache.Redis/Handlers/U/Set.cs
// -----------------------------------------------------------------------
// <copyright file="Set.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Contracts.DistributedCache.Redis.Handlers.U;

using System.Net;
using System.Text.Json;
using D2.Contracts.Handler;
using D2.Contracts.Result;
using D2.Contracts.Utilities.Serialization;
using Google.Protobuf;
using Microsoft.Extensions.Logging;
using StackExchange.Redis;
using S = D2.Contracts.Interfaces.Caching.Distributed.Handlers.U.IUpdate;

/// <summary>
/// Handler for setting a value in the distributed cache.
/// </summary>
///
/// <typeparam name="TValue">
/// The type of the value to cache.
/// </typeparam>
public class Set<TValue> : BaseHandler<
        S.ISetHandler<TValue>, S.SetInput<TValue>, S.SetOutput>,
    S.ISetHandler<TValue>
{
    private readonly IConnectionMultiplexer r_redis;

    /// <summary>
    /// Initializes a new instance of the <see cref="Set{TValue}"/> class.
    /// </summary>
    ///
    /// <param name="redis">
    /// The Redis connection multiplexer.
    /// </param>
    /// <param name="context">
    /// The handler context.
    /// </param>
    public Set(
        IConnectionMultiplexer redis,
        IHandlerContext context)
        : base(context)
    {
        r_redis = redis;
    }

    /// <inheritdoc/>
    protected override async ValueTask<D2Result<S.SetOutput?>> ExecuteAsync(
        S.SetInput<TValue> input,
        CancellationToken ct = default)
    {
        try
        {
            // Serialize the value.
            var bytes = input.Value is IMessage message
                ? message.ToByteArray()
                : JsonSerializer.SerializeToUtf8Bytes(
                    input.Value,
                    SerializerOptions.SR_IgnoreCycles);

            // Connect to Redis and set the value.
            var db = r_redis.GetDatabase();
            await db.StringSetAsync(input.Key, bytes, input.Expiration);

            // Return success.
            return D2Result<S.SetOutput?>.Ok(
                new S.SetOutput(),
                traceId: TraceId);
        }
        catch (RedisException ex)
        {
            Context.Logger.LogError(
                ex,
                "RedisException occurred while setting value for key '{Key}'. TraceId: {TraceId}",
                input.Key,
                TraceId);

            return D2Result<S.SetOutput?>.Fail(
                ["Unable to connect to Redis."],
                HttpStatusCode.ServiceUnavailable,
                errorCode: ErrorCodes.SERVICE_UNAVAILABLE,
                traceId: TraceId);
        }
        catch (JsonException ex)
        {
            Context.Logger.LogError(
                ex,
                "JsonException occurred while serializing value for key '{Key}'. TraceId: {TraceId}",
                input.Key,
                TraceId);

            const string err_msg = "Value could not be serialized.";
            return D2Result<S.SetOutput?>.Fail(
                [err_msg],
                HttpStatusCode.InternalServerError,
                [[nameof(S.SetInput<TValue>.Value), err_msg]],
                ErrorCodes.COULD_NOT_BE_SERIALIZED,
                TraceId);
        }

        // Let the base handler catch any other exceptions.
    }
}

File: backends/node/shared/implementations/caching/distributed/redis/src/handlers/u/set.ts
import type Redis from "ioredis";
import { BaseHandler, type IHandlerContext } from "@d2/handler";
import { D2Result, ErrorCodes, HttpStatusCode } from "@d2/result";
import type { DistributedCache } from "@d2/interfaces";
import { JsonCacheSerializer, type ICacheSerializer } from "../../serialization.js";

export class Set<TValue>
  extends BaseHandler<DistributedCache.SetInput<TValue>, DistributedCache.SetOutput>
  implements DistributedCache.ISetHandler<TValue>
{
  private readonly redis: Redis;
  private readonly serializer: ICacheSerializer<TValue>;

  constructor(redis: Redis, context: IHandlerContext, serializer?: ICacheSerializer<TValue>) {
    super(context);
    this.redis = redis;
    this.serializer = serializer ?? new JsonCacheSerializer<TValue>();
  }

  protected async executeAsync(
    input: DistributedCache.SetInput<TValue>,
  ): Promise<D2Result<DistributedCache.SetOutput | undefined>> {
    try {
      let serialized: string | Buffer;
      try {
        serialized = this.serializer.serialize(input.value);
      } catch {
        return D2Result.fail({
          messages: ["Value could not be serialized."],
          statusCode: HttpStatusCode.InternalServerError,
          errorCode: ErrorCodes.COULD_NOT_BE_SERIALIZED,
          traceId: this.traceId,
        });
      }

      if (input.expirationMs !== undefined) {
        await this.redis.set(input.key, serialized, "PX", input.expirationMs);
      } else {
        await this.redis.set(input.key, serialized);
      }

      return D2Result.ok({ data: {}, traceId: this.traceId });
    } catch {
      return D2Result.fail({
        messages: ["Unable to connect to Redis."],
        statusCode: HttpStatusCode.ServiceUnavailable,
        errorCode: ErrorCodes.SERVICE_UNAVAILABLE,
        traceId: this.traceId,
      });
    }
  }
}
